### PR TITLE
#58 - abstract type hierarchy

### PR DESCRIPTION
**Pull request by schillic**
done:
- added numeric type to `ZeroSet`
- added abstract type hierarchy
- added all missing functions (except #5 and #84)
- added unit tests for interface functions
- updated documentation

### CODE REVIEW

File: src/VPolygon.jl [hunk 1]
```diff
@@ -39,23 +37,69 @@ struct VPolygon{N<:Real} <: LazySet
...
-function dim(P::VPolygon)::Int
-    return 2
+function tovrep(P::VPolygon{N})::VPolygon{N} where {N<:Real}
+    return P
 end
 
+"""
+    tohrep(P::VPolygon{N})::HPolygonal{N} where {N<:Real}
+
+Build a contraint representation of the given polygon.
```
Comment (mforets) [inline src/VPolygon.jl]: contraint -> constraint

File: docs/src/lib/representations.md [hunk 2]
```diff
@@ -142,7 +142,7 @@ dim(::EmptySet)
 ZeroSet
 dim(::ZeroSet)
 σ(::AbstractVector{Float64}, ::ZeroSet)
-∈(::AbstractVector{Float64}, ::ZeroSet)
+∈(::AbstractVector{Float64}, ::ZeroSet{Float64})
```
Comment (mforets) [inline docs/src/lib/representations.md]: let's merge this unrelated change via another branch? one question: don't you need this annotation also in the previous line, `σ(::AbstractVector{Float64}, ::ZeroSet)`?

**Comment by mforets** [comment]
could you consider splitting this big PR in smaller chunks? we can incorporate each new abstract type in stages.

**Comment by schillic** [comment]
>  could you consider splitting this big PR in smaller chunks? we can incorporate each new abstract type in stages.

Do the intermediate commits have to compile? It's rather complicated otherwise.

**Comment by mforets** [comment]
maybe useful is to use the git cherry-pick from a new branch for one or more commits.

File: src/VPolygon.jl (hunk 1 above)
Comment (schillic) [reply src/VPolygon.jl]: fixed, thx

**Comment by schillic** [comment]
What would theses chunks look like?

**Comment by schillic** [comment]
From my side this PR is ready.

@mforets: If you tell me what you have in mind, I can try the splitting, but I do not see a big benefit. I mostly reordered functions and copy-pasted them around, no new features really.

**Comment by schillic** [comment]
I reorganized the commits a bit, is this acceptable?

**Comment by schillic** [comment]
I replaced `radius_b` by `radius_hyperrectangle` as discussed. It always returns a vector now for consistency.

**Comment by mforets** [comment]
good! do merge when you are fine with the change 👍
